app/src/main/java/com/jalasoft/routesapp/ui/auth/login/view/LoginFragment.kt

```diff
@@ -1,32 +1,208 @@
+
+                override fun onError(error: FacebookException) {
+                    Log.d("Facebook", "facebook:onError", error)
+                }
+            }
+        )
+    }
+
+    private fun googleConfiguration() {
+        val gso = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
+            .requestIdToken(getString(R.string.default_web_cliente_id))
+            .requestEmail()
+            .build()
+        googleSingInClient = GoogleSignIn.getClient(binding.root.context, gso)
+    }
+
+    private fun loginWithEmailAndPassword() {
+        showProgress(true)
+        val email = binding.loginEmail.text.toString()
+        val password = binding.loginPassword.text.toString()
+
+        viewModel.loginUserAuth(email, password)
+    }
+
+    private fun observers() {
+        val errorObserver = Observer<String> { errorMessage ->
+            binding.txtError.isVisible = true
+            binding.txtError.text = errorMessage
+            showProgress(false)
+        }
+        val resultObserver = Observer<Boolean> {
+            showProgress(false)
+            binding.loginEmail.setText("")
+            binding.loginPassword.setText("")
+            findNavController().navigate(R.id.homeFragment)
+            Toast.makeText(context, "Login Successfully", Toast.LENGTH_SHORT).show()
+        }
+        val googleAndFacebookObserver = Observer<Boolean> { value ->
+            if (value) {
+                findNavController().navigate(R.id.homeFragment)
+                showProgress(false)
+            }
+        }
+        viewModel.errorMessage.observe(this, errorObserver)
+        viewModel.loginIsSuccessful.observe(this, resultObserver)
+        viewModel.signInGoogleOrFacebook.observe(this, googleAndFacebookObserver)
+    }
+
+    private fun showProgress(show: Boolean) {
+        if (show) {
+            binding.pbLogin.visibility = View.VISIBLE
+        } else {
+            binding.pbLogin.visibility = View.GONE
+        }
+    }
+
+    private fun signInGoogle() {
+        val signInIntent = googleSingInClient.signInIntent
+        launcher.launch(signInIntent)
+    }
+
+    private val launcher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
+            result ->
+        if (result.resultCode == Activity.RESULT_OK) {
+            val task = GoogleSignIn.getSignedInAccountFromIntent(result.data)
+            handleGoogleResults(task)
+        }
+    }
+
+    private fun handleGoogleResults(task: Task<GoogleSignInAccount>) {
+        val account: GoogleSignInAccount? = task.result
+        if (account != null) {
+            val credential = GoogleAuthProvider.getCredential(account.idToken, null)
+            updateUI(account.displayName.toString(), account.email.toString(), UserTypeLogin.GOOGLE, credential)
+        }
+    }
+
+    private fun handleFacebookAccessToken(token: AccessToken) {
+        val credential = FacebookAuthProvider.getCredential(token.token)
```

Comment:
This call is the same on both fragments, shouldn't it be handled on a Manager / Repository?. Remember Single responsibility principle.

Reply:
Solved